File: libs/ktheme-swift/Sources/Ktheme/Model/Theme.swift
// libs/ktheme-swift/Sources/Ktheme/Model/Theme.swift
//
// Top-level Theme type — Codable, mirrors `themes/examples/*.json` 1:1.
// Field names match the JSON exactly so default Codable just works.

import Foundation

public struct Theme: Codable, Hashable, Sendable {
    public let metadata: ThemeMetadata
    public let darkMode: Bool
    public let colorScheme: ColorScheme
    public let effects: Effects
    public let typography: Typography?
    public let adaptation: Adaptation?

    public init(
        metadata: ThemeMetadata,
        darkMode: Bool = true,
        colorScheme: ColorScheme,
        effects: Effects = Effects(),
        typography: Typography? = nil,
        adaptation: Adaptation? = nil
    ) {
        self.metadata = metadata
        self.darkMode = darkMode
        self.colorScheme = colorScheme
        self.effects = effects
        self.typography = typography
        self.adaptation = adaptation
    }
}

public struct ThemeMetadata: Codable, Hashable, Sendable {
    public let id: String
    public let name: String
    public let description: String
    public let author: String
    public let version: String
    public let tags: [String]
    public let createdAt: String?
    public let updatedAt: String?

    public init(id: String, name: String, description: String = "",
                author: String = "Ktheme", version: String = "1.0.0",
                tags: [String] = [], createdAt: String? = nil, updatedAt: String? = nil) {
        self.id = id; self.name = name; self.description = description
        self.author = author; self.version = version; self.tags = tags
        self.createdAt = createdAt; self.updatedAt = updatedAt
    }
}

/// Material Design 3 color roles. Hex strings (#RRGGBB or #AARRGGBB).
public struct ColorScheme: Codable, Hashable, Sendable {
    public let primary, onPrimary, primaryContainer, onPrimaryContainer: String
    public let secondary, onSecondary, secondaryContainer, onSecondaryContainer: String
    public let tertiary, onTertiary, tertiaryContainer, onTertiaryContainer: String
    public let error, onError, errorContainer, onErrorContainer: String
    public let background, onBackground: String
    public let surface, onSurface, surfaceVariant, onSurfaceVariant: String
    public let outline, outlineVariant: String
    public let scrim: String
    public let inverseSurface, inverseOnSurface, inversePrimary: String
    public let stateLayers: StateLayers?
    public let semanticRoles: SemanticRoles?
}

public struct StateLayers: Codable, Hashable, Sendable {
    public let hover, pressed, focused, dragged: String
}

public struct SemanticRoles: Codable, Hashable, Sendable {
    public let success, warning, info, critical: String
}

public struct Typography: Codable, Hashable, Sendable {
    public let fontFamily: String
    public let fontSize: FontSize
    public let fontWeight: FontWeight
    public let lineHeight: Double
    public let letterSpacing: Double
}
public struct FontSize: Codable, Hashable, Sendable {
    public let small, medium, large, xlarge: Int
}
public struct FontWeight: Codable, Hashable, Sendable {
    public let light, regular, medium, bold: Int
}

public struct Adaptation: Codable, Hashable, Sendable {
    public let layout: LayoutAdaptation?
    public let icons: IconAdaptation?
    public let componentOverrides: [ComponentOverride]?
}

public struct LayoutAdaptation: Codable, Hashable, Sendable {
    public let density: Density
    public let cornerStyle: CornerStyle
    public let spacingScale: Double
    public let panelStyle: PanelStyle
    public let navigationStyle: NavigationStyle
}

public enum Density: String, Codable, Hashable, Sendable {
    case compact, comfortable, spacious
    public var scale: Double {
        switch self { case .compact: 0.92; case .comfortable: 1.0; case .spacious: 1.25 }
    }
}

public enum CornerStyle: String, Codable, Hashable, Sendable {
    case sharp, rounded, pill
    public var radius: CGFloat {
        switch self { case .sharp: 0; case .rounded: 12; case .pill: 24 }
    }
}

public enum PanelStyle: String, Codable, Hashable, Sendable {
    case flat, glass, elevated
}

public enum NavigationStyle: String, Codable, Hashable, Sendable {
    case tabs, rail, drawer, pivot, bottom
}

public struct IconAdaptation: Codable, Hashable, Sendable {
    public let family: IconFamily
    public let style: IconStyle
    public let sizeScale: Double
    public let strokeWidth: Double
    public let cornerStyle: CornerStyle
}
public enum IconFamily: String, Codable, Hashable, Sendable {
    case material, fluent, sf, line, duotone, custom
}
public enum IconStyle: String, Codable, Hashable, Sendable {
    case line, filled, duotone
}

/// Component-level CSS-style override targeted at a selector. Carried
/// across as raw values (string or number); SwiftUI consumers map the
/// subset they care about (border-radius, padding, text-transform).
public struct ComponentOverride: Codable, Hashable, Sendable {
    public let selector: String
    public let styles: [String: StyleValue]
}

/// Either a string ("uppercase") or a number (16) inside a styles map.
public enum StyleValue: Codable, Hashable, Sendable {
    case string(String)
    case number(Double)

    public init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if let s = try? c.decode(String.self) { self = .string(s); return }
        if let d = try? c.decode(Double.self) { self = .number(d); return }
        throw DecodingError.dataCorruptedError(in: c,
            debugDescription: "Expected string or number for StyleValue")
    }
    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .string(let s): try c.encode(s)
        case .number(let d): try c.encode(d)
        }
    }
    public var stringValue: String? { if case .string(let s) = self { s } else { nil } }
    public var numberValue: Double? { if case .number(let d) = self { d } else { nil } }
}

File: libs/ktheme-core/src/main/kotlin/io/ktheme/model/Theme.kt
// libs/ktheme-core/src/main/kotlin/io/ktheme/model/Theme.kt
//
// Top-level Theme data class — 1:1 with the canonical Ktheme JSON shape
// shipped in `themes/examples/*.json`. Field names match the JSON exactly
// so kotlinx.serialization can round-trip without custom serializers.
//
// Anything optional in the JSON is nullable here so older preset files
// (which omit `adaptation`, `typography`, etc.) parse cleanly.

package io.ktheme.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
public data class Theme(
    public val metadata: ThemeMetadata,
    public val darkMode: Boolean = true,
    public val colorScheme: ColorScheme,
    public val effects: Effects = Effects(),
    public val typography: Typography? = null,
    public val adaptation: Adaptation? = null,
)

@Serializable
public data class ThemeMetadata(
    public val id: String,
    public val name: String,
    public val description: String = "",
    public val author: String = "Ktheme",
    public val version: String = "1.0.0",
    public val tags: List<String> = emptyList(),
    public val createdAt: String? = null,
    public val updatedAt: String? = null,
)

/**
 * Material Design 3 color roles. Hex strings (#RRGGBB or #AARRGGBB).
 * Use [Theme.toM3] in the Compose integration to convert to androidx
 * `ColorScheme`.
 */
@Serializable
public data class ColorScheme(
    public val primary: String,
    public val onPrimary: String,
    public val primaryContainer: String,
    public val onPrimaryContainer: String,
    public val secondary: String,
    public val onSecondary: String,
    public val secondaryContainer: String,
    public val onSecondaryContainer: String,
    public val tertiary: String,
    public val onTertiary: String,
    public val tertiaryContainer: String,
    public val onTertiaryContainer: String,
    public val error: String,
    public val onError: String,
    public val errorContainer: String,
    public val onErrorContainer: String,
    public val background: String,
    public val onBackground: String,
    public val surface: String,
    public val onSurface: String,
    public val surfaceVariant: String,
    public val onSurfaceVariant: String,
    public val outline: String,
    public val outlineVariant: String,
    public val scrim: String = "#000000",
    public val inverseSurface: String,
    public val inverseOnSurface: String,
    public val inversePrimary: String,
    /** Optional MD3 state-layer tints. */
    public val stateLayers: StateLayers? = null,
    /** Optional product-semantic roles (success/warning/info/critical). */
    public val semanticRoles: SemanticRoles? = null,
)

@Serializable
public data class StateLayers(
    public val hover: String,
    public val pressed: String,
    public val focused: String,
    public val dragged: String,
)

@Serializable
public data class SemanticRoles(
    public val success: String,
    public val warning: String,
    public val info: String,
    public val critical: String,
)

@Serializable
public data class Typography(
    public val fontFamily: String = "system-ui, -apple-system, sans-serif",
    public val fontSize: FontSize = FontSize(),
    public val fontWeight: FontWeight = FontWeight(),
    public val lineHeight: Double = 1.5,
    public val letterSpacing: Double = 0.0,
)

@Serializable
public data class FontSize(
    public val small: Int = 12,
    public val medium: Int = 16,
    public val large: Int = 20,
    public val xlarge: Int = 28,
)

@Serializable
public data class FontWeight(
    public val light: Int = 300,
    public val regular: Int = 400,
    public val medium: Int = 500,
    public val bold: Int = 700,
)

@Serializable
public data class Adaptation(
    public val layout: LayoutAdaptation? = null,
    public val icons: IconAdaptation? = null,
    public val componentOverrides: List<ComponentOverride> = emptyList(),
)

@Serializable
public data class LayoutAdaptation(
    public val density: Density = Density.Comfortable,
    public val cornerStyle: CornerStyle = CornerStyle.Rounded,
    public val spacingScale: Double = 1.0,
    public val panelStyle: PanelStyle = PanelStyle.Flat,
    public val navigationStyle: NavigationStyle = NavigationStyle.Tabs,
)

@Serializable
public enum class Density {
    @SerialName("compact") Compact,
    @SerialName("comfortable") Comfortable,
    @SerialName("spacious") Spacious;

    public val scale: Double get() = when (this) {
        Compact -> 0.92
        Comfortable -> 1.0
        Spacious -> 1.25
    }
}

@Serializable
public enum class CornerStyle {
    @SerialName("sharp") Sharp,
    @SerialName("rounded") Rounded,
    @SerialName("pill") Pill;

    /** Default corner radius in dp/pt for this style. */
    public val radiusDp: Int get() = when (this) {
        Sharp -> 0
        Rounded -> 12
        Pill -> 24
    }
}

@Serializable
public enum class PanelStyle {
    @SerialName("flat") Flat,
    @SerialName("glass") Glass,
    @SerialName("elevated") Elevated,
}

@Serializable
public enum class NavigationStyle {
    @SerialName("tabs") Tabs,
    @SerialName("rail") Rail,
    @SerialName("drawer") Drawer,
    @SerialName("pivot") Pivot,
    @SerialName("bottom") Bottom,
}

@Serializable
public data class IconAdaptation(
    public val family: IconFamily = IconFamily.Material,
    public val style: IconStyle = IconStyle.Line,
    public val sizeScale: Double = 1.0,
    public val strokeWidth: Double = 1.5,
    public val cornerStyle: CornerStyle = CornerStyle.Rounded,
)

@Serializable
public enum class IconFamily {
    @SerialName("material") Material,
    @SerialName("fluent")   Fluent,
    @SerialName("sf")       Sf,
    @SerialName("line")     Line,
    @SerialName("duotone")  Duotone,
    @SerialName("custom")   Custom,
}

@Serializable
public enum class IconStyle {
    @SerialName("line")    Line,
    @SerialName("filled")  Filled,
    @SerialName("duotone") Duotone,
}

/**
 * A free-form CSS-style override targeted at a selector. Kept as a string
 * map so any platform can interpret a useful subset (Compose ignores
 * web-only props like `backdrop-filter`; SwiftUI maps `border-radius`
 * to corner radius, etc.).
 */
@Serializable
public data class ComponentOverride(
    public val selector: String,
    public val styles: Map<String, JsonStringOrNumber> = emptyMap(),
)

/**
 * Theme JSON freely mixes string and numeric values inside `styles`
 * (e.g. `"padding-inline": 16` and `"text-transform": "uppercase"`).
 * This wrapper keeps them legible without forcing every consumer to
 * parse JsonElement directly.
 */
@Serializable(with = JsonStringOrNumberSerializer::class)
public data class JsonStringOrNumber(public val raw: String) {
    public val asDoubleOrNull: Double? get() = raw.toDoubleOrNull()
    override fun toString(): String = raw
}
